### DISCUSSION TITLE
C# 10: Adding a field initializer to a struct type 'breaks' it

### DISCUSSION THREAD

**Discussion by pkulikov** [dotnet/csharplang#5148] [Q&A]
First of all, the C# 9-compliant code that doesn't compile (CS0843: Auto-implemented property 'Program.Example.A' must be fully assigned before control is returned to the caller):

```csharp
    public readonly struct Example
    {
        public Example(int a)
        {  }

        public int A { get; }
        public int B { get; }
    }
```

To fix the build error, let's zero all the fields with the `this()` call:

```csharp
    public readonly struct Example
    {
        public Example(int a) : this()
        {  }

        public int A { get; }
        public int B { get; }
    }
```

However, when, in C# 10, I add the field initializer, the code produces the same error as the first example (CS0843: Auto-implemented property 'Program.Example.A' must be fully assigned before control is returned to the caller):

```csharp
    public readonly struct Example
    {
        public Example(int a) : this()
        {  }

        public int A { get; }
        public int B { get; } = 10;
    }
```

That doesn't look consistent: if the second example is fine, why the third one fails to compile? Is `: this()` part ignored by the compiler? I would expect all fields to be zeroed and then `B` set to `10`.

**Comment by svick** [comment]
> I would expect all fields to be zeroed and then `B` set to `10`.

Why is this the expected behavior? I'd say the closest equivalent code for a class is:

```c#
public class Example
{
    public Example()
    {
        A = 0;
        B = 0;
    }

    public Example(int a) : this()
    {  }

    public int A { get; }
    public int B { get; } = 10;
}
```

This code executes in the opposite order: `B` is first set to 10 and only after that are both properties set to 0.

**Reply by pkulikov** [reply]
@svick thank you, I agree with you; indeed - initializers go first before any constructor call.

Then the question remains: why does the struct example stop to compile instead of behaving like your class example?

Furthermore, there are still some differences between class and struct. The following example compiles:

```csharp
public readonly struct Example
    {
        public Example(int a) : this()
        {
            A = 4;
        }

        public int A { get; }
        public int B { get; } = 10;
    }
```

and any call to `Example(int a)` produces an instance with `B` set to `10`. In case of the class example, `B` would be set to `0` at the end. 

Should the compiler warn in the preceding example about 'useless' `this()` call?

And at last, for

```csharp
    public readonly struct Example
    {
        public Example()
        {
            A = 0;
            B = 0;
        }

        public Example(int a) : this()
        {
            A = 4;
        }

        public int A { get; }
        public int B { get; } = 10;
    }
```
any call to `Example(int a)` produces an instance with `B` zeroed.

In general, it may be a bit hard to reason about the `: this()` call in the struct context.

**Comment by TonyValenti** [comment]
@pkulikov - At the company I work for, In order to make intent clear, we don't allow a property/field to be initialized from a constructor and a default value. Initializing it through multiple methods makes understanding intent confusing as you see here.
